akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -408,36 +432,53 @@ final case class Intersperse[T](start: Option[T], inject: T, end: Option[T]) ext
 /**
  * INTERNAL API
  */
-private[akka] final case class Grouped[T](n: Int) extends PushPullStage[T, immutable.Seq[T]] {
-  private val buf = {
-    val b = Vector.newBuilder[T]
-    b.sizeHint(n)
-    b
-  }
-  private var left = n
+private[akka] final case class Grouped[T](n: Int) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
+  require(n > 0, "n must be greater than 0")
 
-  override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
-    buf += elem
-    left -= 1
-    if (left == 0) {
-      val emit = buf.result()
-      buf.clear()
-      left = n
-      ctx.push(emit)
-    } else ctx.pull()
-  }
+  val in = Inlet[T]("Grouped.in")
+  val out = Outlet[immutable.Seq[T]]("Grouped.out")
+  override val shape: FlowShape[T, immutable.Seq[T]] = FlowShape(in, out)
 
-  override def onPull(ctx: Context[immutable.Seq[T]]): SyncDirective =
-    if (ctx.isFinishing) {
-      val elem = buf.result()
-      buf.clear()
-      left = n
-      ctx.pushAndFinish(elem)
-    } else ctx.pull()
+  override protected val initialAttributes: Attributes = DefaultAttributes.grouped
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+    private val buf = {
+      val b = Vector.newBuilder[T]
+      b.sizeHint(n)
+      b
+    }
+    var left = n
+
+    override def onPush(): Unit = {
+      buf += grab(in)
+      left -= 1
+      if (left == 0) {
+        val elements = buf.result()
+        buf.clear()
+        left = n
+        push(out, elements)
+      } else {
+        pull(in)
+      }
+    }
+
+    override def onPull(): Unit = {
+      pull(in)
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (isAvailable(out) && left < n) {
+        val elements = buf.result()
+        buf.clear()
+        left = n
+        push(out, elements)
```

Comment:
So, if there are elements in the buffer but downstream isn't ready this will loose elements, please go over this again and reintroduce the original behavior of deferring completion of this stage until there is demand.